File: BusCatalog.Api/BusCatalog.Api/Domain/HealthCheck/HealthCheckAdapter.cs
using System.Net;
using BusCatalog.Api.Domain.HealthCheck.Ports;
using static BusCatalog.Api.Domain.HealthCheck.Messages.ServiceMessages;

namespace BusCatalog.Api.Domain.HealthCheck;

public interface IHealthCheckAdapter
{
    Task<HealthCheckResponse> SendAsync(string url, string application);
}

public sealed class HealthCheckAdapter : IHealthCheckAdapter
{
    private readonly HttpClient _http;
    private readonly ILogger<HealthCheckAdapter> _logger;
    private readonly static int InvalidStatusCode = (int)HttpStatusCode.ServiceUnavailable;

    public HealthCheckAdapter(
        HttpClient http,
        ILogger<HealthCheckAdapter> logger)
    {
        _http = http;
        _logger = logger;
    }

    public async Task<HealthCheckResponse> SendAsync(string url, string application)
    {
        try
        {
            return await HandleSuccess(url, application);
        }
        catch (HttpRequestException)
        {
            return HandleError(url, application);
        }
    }

    private async Task<HealthCheckResponse> HandleSuccess(string url, string application)
    {
        var response = await _http.SendAsync(new HttpRequestMessage(HttpMethod.Get, url));
        _logger.LogInformation(HealthCheckResult, application, url, (int)response.StatusCode);

        return HealthCheckResponse.Generate(
            url,
            (int)response.StatusCode,
            response.IsSuccessStatusCode);
    }

    private HealthCheckResponse HandleError(string url, string application)
    {
        _logger.LogWarning(HealthCheckResult, application, url, InvalidStatusCode);

        return HealthCheckResponse.Generate(
            url,
            InvalidStatusCode,
            success: false);
    }
}

File: BusCatalog.Scraper/src/main/java/br/com/marcomvidal/buscatalog/scraper/healthcheck/HealthCheckAdapter.java
package br.com.marcomvidal.buscatalog.scraper.healthcheck;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import br.com.marcomvidal.buscatalog.scraper.healthcheck.messages.ServiceMessages;
import br.com.marcomvidal.buscatalog.scraper.healthcheck.ports.HealthCheckResponse;

@Service
public class HealthCheckAdapter {
    private final RestClient http;
    private final Logger logger = LoggerFactory.getLogger(HealthCheckAdapter.class);
    private final int INVALID_STATUS_CODE = HttpStatus.SERVICE_UNAVAILABLE.value();

    public HealthCheckAdapter(RestClient.Builder http) {
        this.http = http.build();
    }

    public HealthCheckResponse send(String url, String application) {
        try {
            return handleSuccess(url, application);
        } catch (Exception ex) {
            return handleError(url, application);
        }
    }

    private HealthCheckResponse handleSuccess(String url, String application) {
        var response = http.get().uri(url).retrieve().toBodilessEntity();
            
        var statusCode = response.getStatusCode();

        logger.info(
            ServiceMessages.HEALTH_CHECK_RESULT.getMessage(),
            application,
            url,
            statusCode);

        return HealthCheckResponse.generate(
            url,
            statusCode.value(),
            statusCode.is2xxSuccessful());
    }

    private HealthCheckResponse handleError(String url, String application) {
        logger.warn(
            ServiceMessages.HEALTH_CHECK_RESULT.getMessage(),
            application,
            url,
            INVALID_STATUS_CODE);

        return HealthCheckResponse.generate(url, INVALID_STATUS_CODE, false);
    }
}
